common/src/group_by.rs

```diff
@@ -0,0 +1,166 @@
+use std::cell::RefCell;
+use std::iter::Peekable;
+use std::rc::Rc;
+
+pub trait GroupByIteratorExtended: Iterator {
+    /// Return an `Iterator` that groups iterator elements. Consecutive elements that map to the
+    /// same key are assigned to the same group.
+    ///
+    /// The returned Iterator item is `(K, impl Iterator)`, where Iterator are the items of the
+    /// group.
+    ///
+    /// ```
+    /// use tantivy_common::GroupByIteratorExtended;
+    ///
+    /// // group data into blocks of larger than zero or not.
+    /// let data: Vec<i32> = vec![1, 3, -2, -2, 1, 0, 1, 2];
+    /// // groups:               |---->|------>|--------->|
+    ///
+    /// let mut data_grouped = Vec::new();
+    /// // Note: group is an iterator
+    /// for (key, group) in data.into_iter().group_by(|val| *val >= 0) {
+    ///     data_grouped.push((key, group.collect()));
+    /// }
+    /// assert_eq!(data_grouped, vec![(true, vec![1, 3]), (false, vec![-2, -2]), (true, vec![1, 0, 1, 2])]);
+    /// ```
+    fn group_by<K, F>(self, key: F) -> GroupByIterator<Self, F, K>
+    where
+        Self: Sized,
+        F: FnMut(&Self::Item) -> K,
+        K: PartialEq + Copy,
+        Self::Item: Copy,
+    {
+        GroupByIterator::new(self, key)
+    }
+}
+impl<I: Iterator> GroupByIteratorExtended for I {}
+
+pub struct GroupByIterator<I, F, K: Copy>
+where
+    I: Iterator,
+    F: FnMut(&I::Item) -> K,
+{
+    // I really would like to avoid the Rc<RefCell>, but the Iterator is shared between
+    // `GroupByIterator` and `GroupIter`. In practice they are used consecutive and
+    // `GroupByIter` is finished before calling next on `GroupByIterator`. I'm not sure there
+    // is a solution with lifetimes for that, because we would need to enforce it in the usage
+    // somehow.
+    //
+    // One potential solution would be to replace the iterator approach with something similar.
+    inner: Rc<RefCell<GroupByShared<I, F, K>>>,
```

Comment:
This might be a use case for GAT + `LendingIterator` maybe? But that's not an `Iterator` unfortunately.